internal/controller/controlplane/status.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sort"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/collections"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	cpv1beta1 "github.com/k0sproject/k0smotron/api/controlplane/v1beta1"
+	"github.com/k0sproject/version"
+)
+
+// versionMatches checks if the machine version matches the kcp version taking the possibly missing suffix into account
+func versionMatches(machine *clusterv1.Machine, ver string) bool {
+
+	if machine.Spec.Version == nil || *machine.Spec.Version == "" {
+		return false
+	}
+
+	if *machine.Spec.Version == ver {
+		return true
+	}
+
+	machineVersion := *machine.Spec.Version
+	kcpVersion := ver
+
+	// If either of the versions is missing the suffix, we need to add it
+	// But take the suffix from kcp version if present
+	kcpSuffix := getVersionSuffix(kcpVersion)
+	if kcpSuffix == "" {
+		kcpVersion = kcpVersion + "+k0s.0"
+		kcpSuffix = "k0s.0"
```

Comment:
```suggestion
                kcpSuffix = "k0s.0"
		kcpVersion = kcpVersion + kcpSuffix
```
nit: I think less error prone